### ISSUE TITLE
GKE Pod/Node Exhaustion Resulting in Failed Workflows

### ISSUE THREAD

**Issue by mwarnerblu** [argoproj/argo-workflows#4794]
## Summary 

We're evaluating utilizing Argo (Workflows and Events) to replace our current utilization of Airflow.  We have some fairly high volume requirements, starting at around 100 Workflows/minute and going up from there sourced from GCP Pub/Sub.  In general Argo keeps up well processing simple workflows as you've seen before and is much faster than Airflow at scale for task running and handling. 

However, we are seeing issues that appear to be related to the nodes maxing out their pods (110 per node in GKE) and hitting a hard failure state.  I've done some testing with parallelism, retry strategies, and synchronizing workflows to avoid over-saturating nodes during spikes but we've been unable to avoid the hard failure in 5-6% of Workflow runs.  This is partly happening due to the lack of flow control/backpressure within the [GCP Pub Sub](https://github.com/argoproj/argo-events/blob/master/eventsources/sources/gcppubsub/start.go) event source, we're looking at making some changing internally to meet GCP recommendations for [Go Pub/Sub subscription flow control on this as well](https://cloud.google.com/pubsub/docs/samples/pubsub-subscriber-flow-settings#pubsub_subscriber_flow_settings-go) - this is more relevant for the argo-events repo however.

In either case, due to the spiky-nature of Pub/Sub a spike of a couple hundred to thousand will immediately result in a failure state as the nodes have not yet scaled up.  This results in a failure of all workflows until the nodepool scales up to the needed utilization.  While better backpressure on the GCP Pub Sub event source will help resolve the available node/pod sets available for the Workflows, the ability to ensure that those that fail due to an unchangeable (at this point) quota by GKE would be ideal.

I've added a quick and dirty python snippet[1] that can be used to trigger this, in it's current state of shipping 1200 messages immediately you should see a ~5% failure rate across a 4 nodes GKE cluster as it scales up to 6-12 nodes to accommodate the utilization spike. Enforcing backpressure at the GCP Pub Sub event source is likely the ideal way to solve this, but, these failures on Workflows should ideally be retried as the quota issue should be resolved by either the processing of existing workloads or additional nodes being in place.

I know this has been discussed before in [3217](https://github.com/argoproj/argo/issues/3217), we are above that issue in version so it seems to still be cropping up.  If we're doing something improperly here I'm happy to make changes on our side as well!  

Thank you!

## Diagnostics

What Kubernetes provider are you using? 
Google GKE on 1.17.13-gke.2001 on a n2d-standard-2 nodepool with minimum of 4 nodes using Container-Optimized OS that autoscale based on optimizing utilization. 

What version of Argo Workflows are you running? 
```
argo -n argo-events version
argo: v2.12.0-rc5
  BuildDate: 2020-12-10T13:23:30Z
  GitCommit: 3aa86fffb7c975e3a39302f5b2e37f99fe58fa4f
  GitTreeState: clean
  GitTag: v2.12.0-rc5
  GoVersion: go1.13
  Compiler: gc
  Platform: linux/amd64
```

**Event Source YAML**
```yaml
apiVersion: argoproj.io/v1alpha1
kind: EventSource
metadata:
  name: gcp-pubsub
spec:
  pubSub:
    first-workload:
      jsonBody: true
      projectID: blu-dev
      topic: blu-dev-argo-incoming
      subscriptionID: argo-events
      credentialSecret:
        name: gcp-credentials
        namespace: argo-events
        key: key.json
```

**Event Sensor to Workflow YAML**
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Sensor
metadata:
  name: gcp-pubsub
spec:
  template:
    serviceAccountName: argo-events-sa
  dependencies:
    - name: gcp
      eventSourceName: gcp-pubsub
      eventName: first-workload
  triggers:
    - template:
        name: gcp-workflow
        parallelism: 200
        nodeSelector:
          "node-role.kubernetes.io/argo-spot-worker": "true"
        k8s:
          group: argoproj.io
          version: v1alpha1
          resource: workflows
          operation: create
          source:
            resource:
              apiVersion: argoproj.io/v1alpha1
              kind: Workflow
              metadata:
                generateName: gcp-workflow-
              spec:
                # Synchronize across the threat source
                # - attempting to hirewire lack of P/S flow control
                synchronization:
                  semaphore:
                    configMapKeyRef:
                      name: gcp-pubsub-lock
                      key: workflow
                # must complete in 10m
                activeDeadlineSeconds: 600
                # keep workflows for 1d 86400
                ttlStrategy:
                  secondsAfterCompletion: 1800 
                # delete all pods as soon as they complete
                podGC:
                  strategy: OnPodSuccess  # If you use OnPodComplete retry won't work 
                # Ensure resubmission on pending move to defeat gke issues
                resubmitPendingPods: true
                # Retry Strategy
                retryStrategy:
                  limit: 3
                  retryPolicy: "Always"
                  backoff:
                    duration: "1m"    
                    factor: 2
                    maxDuration: "5m"
                # Definition
                entrypoint: whalesay
                arguments:
                  parameters:
                  - name: message
                    # value will get overridden by the event payload
                    value: hello world
                templates:
                - name: whalesay
                  inputs:
                    parameters:
                    - name: message
                  container:
                    image: docker/whalesay:latest
                    command: [cowsay]
                    args: ["{{inputs.parameters.message}}"]
                    resources:
                      requests:
                        memory: "64Mi"
                        cpu: "150m"
                      limits:
                        memory: "96Mi"
                        cpu: "250m"
          parameters:
            - src:
                dependencyName: gcp
              dest: spec.arguments.parameters.0.value[0]
```

```
time="2020-12-17T21:10:44Z"   level=info msg="Workflow update successful" namespace=argo-events   phase=Error resourceVersion=708382 workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:44Z"   level=info msg="Checking daemoned children of "   namespace=argo-events workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:44Z"   level=info msg="Marking workflow completed" namespace=argo-events   workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:44Z"   level=info msg="Updated message    -> pod deleted" namespace=argo-events   workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:44Z"   level=info msg="Updated phase Running -> Error"   namespace=argo-events workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:44Z"   level=error msg="Pod for node deleted"   displayName=gcp-workflow-tbgpl namespace=argo-events node=gcp-workflow-tbgpl   templateName=whalesay workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:43Z"   level=info msg="Processing workflow" namespace=argo-events   workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:38Z"   level=info msg="Workflow update successful" namespace=argo-events   phase=Running resourceVersion=707995 workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:38Z"   level=info msg="node gcp-workflow-tbgpl message: Pending   2.512673824s" namespace=argo-events workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:38Z"   level=info msg="Mark node gcp-workflow-tbgpl as Pending, due to:   Operation cannot be fulfilled on resourcequotas   \"gke-resource-quotas\": the object has been modified; please apply   your changes to the latest version and try again" namespace=argo-events   workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:35Z"   level=info msg="Pod node gcp-workflow-tbgpl initialized Pending"   namespace=argo-events workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:35Z"   level=info msg="Updated phase    -> Running" namespace=argo-events   workflow=gcp-workflow-tbgpl
time="2020-12-17T21:10:35Z"   level=info msg="Processing workflow" namespace=argo-events   workflow=gcp-workflow-tbgpl
```

## Python PubSub Populator Snipper
- Requires google-cloud-pubsub
```python
#!/usr/bin/env python3

import os
import time
from google.cloud import pubsub_v1

futures = dict()


def get_callback(f, data):
    def callback(f):
        try:
            print(f.result())
            futures.pop(data)
        except:  # noqa
            print("Please handle {} for {}.".format(f.exception(), data))

    return callback


def publish():
    publisher = pubsub_v1.PublisherClient()
    # Vars
    project_id = os.getenv('GOOGLE_CLOUD_PROJECT')
    topic_id = 'blu-dev-argo-incoming'
    # Define full path
    topic_path = publisher.topic_path(project_id, topic_id)
    # Loop and ship total
    for i in range(1200):
        data = str(i)
        futures.update({data: None})
        # When you publish a message, the client returns a future.
        future = publisher.publish(topic_path, data.encode("utf-8"), message='test')
        futures[data] = future
        # Publish failures shall be handled in the callback function.
        future.add_done_callback(get_callback(future, data))
    
    # Wait for all the publish futures to resolve before exiting.
    while futures:
        time.sleep(5)

def go():
    print('Running Pub/Sub Populator for Task Runner') 
    publish()


if __name__ == '__main__':
    go()
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
I'm not clear on what your symptom is.  If it is a `pod deleted` error message. This is addressed in v2.12.

Can you please try the following:

```
    Environment:
      DEFAULT_REQUEUE_TIME:      10s ;# default is 2s
```

**Comment by mwarnerblu** [comment]
Sorry @alexec I forgot to include the argo-details side and just included the logs. We are v2.12.0-rc5 and do appear to be getting the pod delete error message.  However if I'm approaching this improperly I'm happy to rework of course.  Below is the output for a Error'd out workflow as described above.

```
Name:                gcp-workflow-8g2cs
Namespace:           argo-events
ServiceAccount:      default
Status:              Error
Message:             pod deleted
Conditions:          
 Completed           True
Created:             Tue Dec 22 12:46:30 -0500 (31 seconds ago)
Started:             Tue Dec 22 12:46:32 -0500 (29 seconds ago)
Finished:            Tue Dec 22 12:46:38 -0500 (23 seconds ago)
Duration:            6 seconds
Progress:            
Parameters:          
  message:           hello world

STEP                   TEMPLATE  PODNAME             DURATION  MESSAGE
 ⚠ gcp-workflow-8g2cs  whalesay  gcp-workflow-8g2cs  6s        pod deleted
```

I'll try the DEFAULT_REQUEUE_TIME change, thanks!
